game-core/src/main/java/games/strategy/net/nio/ClientQuarantineConversation.java

```diff
@@ -130,6 +132,12 @@ public Action message(final Serializable serializable) {
           }
           localName = strings[0];
           serverName = strings[1];
+          if (strings.length > 2) {
+            passwordChangeRequired =
+                strings[2] != null
+                    && strings[2].equals(ServerQuarantineConversation.CHANGE_PASSWORD);
+          }
```

Comment:
Not blaming you for this, but I have no idea in the slightest what's happening here

Reply:
After successful login the server sends back an array with the client name:

ServerQuarantineConversation.java:128
```
          // send the node its assigned name and our name
          send(new String[] {remoteName, serverMessenger.getLocalNode().getName(), error});
          // send the node its and our address as we see it
          send(
              new InetSocketAddress[] {
                (InetSocketAddress) channel.socket().getRemoteSocketAddress(),
                serverMessenger.getLocalNode().getSocketAddress()
              });
```

New to this update is the third element in the array "error". Typically that will be null, in the change password case it'll be set to the magic string "CHANGE_PASSWORD".

So here we have the client code parsing out the third array element to see if it contains the magic string. Appending to the array avoids having to do deeper changes to the quarantine conversation flow.